server/src/main/java/org/elasticsearch/action/search/BottomSortValuesCollector.java

```diff
@@ -28,9 +31,28 @@ class BottomSortValuesCollector {
     private final FieldComparator<?>[] comparators;
     private final int[] reverseMuls;
 
+    @SuppressWarnings("unused") // updated via TOTAL_HITS_HANDLE
     private volatile long totalHits;
+    @SuppressWarnings("unused") // updated via BOTTOM_SORT_VALUES_HANDLE
     private volatile SearchSortValuesAndFormats bottomSortValues;
 
+    private static final VarHandle TOTAL_HITS_HANDLE;
+    private static final VarHandle BOTTOM_SORT_VALUES_HANDLE;
+
+    static {
+        var lookup = MethodHandles.lookup();
+        try {
+            TOTAL_HITS_HANDLE = lookup.findVarHandle(BottomSortValuesCollector.class, "totalHits", long.class);
+            BOTTOM_SORT_VALUES_HANDLE = lookup.findVarHandle(
+                BottomSortValuesCollector.class,
+                "bottomSortValues",
+                SearchSortValuesAndFormats.class
+            );
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
```

Comment:
I don't feel great about this type of changes. I think that they hinder readability and should be used with care. I wonder what gain it provides in terms of performance and whether it is worthwhile.

Additionally, I am not super confident that we have enough test coverage around this change to refactor it, we should perhaps focus on extending that before refactoring.

Lastly: I would rather focus on a single roundtrip per data node over optimizing all kinds of different places in the search codebase. We can do that later too as needed, but let's get to the core of the change we want to make?

Reply:
As discussed earlier :) This is somewhat orthogonal to the execution change effort. It's mainly an outcome of me learning about the query phase rewriting we do and wanting to make that more obvious and my continued desire to address the many slow transport thread execution warnings we see in ESS logging around this code.

> I wonder what gain it provides in terms of performance and whether it is worthwhile.

It's hard to put a number like x-percent on this obviously. But thing is: the current synchronisation based solution is definitely slower than a CAS solution and much more importantly, **it scales worse and less predictable**.  The memory effects of passing a synchronized block (and thus establishing happens before) are especially unfortunate for us potentially vs. a CAS that doesn't have that effect.
That's the theory of things. In practice for a full run of the http_logs track it appears that `consumeTopDocs` is about half of the CPU time and more than half the wall clock time of `onShardResult` (mind you this is just a single node run with concurrency of 5!). The more CPUs and concurrent threads you run the worse this gets.
=> worthwhile if only to rule out one potential bottle neck on transport threads and get more deterministic behavior.

> Additionally, I am not super confident that we have enough test coverage around this change to refactor it, we should perhaps focus on extending that before refactoring.

It's hard to define coverage here. We have a lot of coverage on sort queries across multiple shards that verify correctness of results. Beyond that we have `BottomSortValuesCollectorTests` that verify that everything's on the up-and-up for single thread uncontended operation I think.
Putting a break-point into the contended CAS case in `server:internalClusterTest` I see we actually hit the contended case where CAS fails quite a bit in these tests -> I think we're good?

>  We can do that later too as needed, but let's get to the core of the change we want to make?

True true, this really was mostly just a random find that I wanted to address as I found it to help transport thread health :)